### ISSUE TITLE
Url Rewriting and Support for http-request set-path

### ISSUE THREAD

**Issue by patran** [vfarcic/docker-flow-proxy#200]
Could this "http-request set-path" be supported along the same lines as "http-request set/add header" ?

External URL: https://myapp.mydomain.com
Back-end server accepts: http://server/myappcontext

http-request set-path could be used to add "/myappcontext". Perhaps this scenario is already supported?

I think could use fe/beTempate, but it's more convenient to not having to deal with a file.

Thanks

**Comment by vfarcic** [comment]
I'm not sure I understood the request. Are you referring to something similar to the [parameters](http://proxy.dockerflow.com/usage/#general-http-query-parameters) `reqPathSearch` and `reqPathReplace`?

**Comment by patran** [comment]
@vfarcic, yes. I ran into reqPathSearch/reqPathReplace a couple of days ago and think they should work. Will try out and report back.

**Comment by ambrons** [comment]
@vfarcic I was trying to follow your example here:
http://proxy.dockerflow.com/swarm-mode-auto/#rewriting-paths

However I'm getting a `BADINDEX` on the service during proxy setup.  Here's the config in stack syntax.  The proxy just keeps reconfiguring until I remove the reqPathSearch and reqPathReplace.

## Config
```
  vote:
    image: dockersamples/examplevotingapp_vote:before
    networks:
      - proxy
      - service
    deploy:
      replicas: 2
      update_config:
        parallelism: 2
      restart_policy:
        condition: on-failure
      labels:
        - com.df.notify=true
        - com.df.distribute=true
        - com.df.xForwardedProto=true
        - com.df.servicePath=/votes
        - com.df.reqPathSearch='/votes/'
        - com.df.reqPathReplace='/'
        - com.df.port=80
```

## Log
```
(BADINDEX) server vote_vote vote_vote:80 check
```

**Comment by vfarcic** [comment]
I used (almost) the same stack and it worked (except that styles were not loaded because they require `static` path to be mapped as well.

Can I assume that you're using the latest DFP version?

Can you post the entire log of the proxy? If that does not give me an indication of the cause of the problem and since I cannot reproduce it on my system, it would be great if we could have a screen-sharing session. If I can take a look at your system I might be able to see what the problem is. For example, the problem might be in networking. Maybe DFP cannot reach the vote service.

**Comment by ambrons** [comment]
Hey @vfarcic thanks for the quick response.

Here's some background.  I'm using Docker for AWS for swarm configuration with current Docker version: 
```
Client:
 Version:      17.05.0-ce-rc1
 API version:  1.28 (downgraded from 1.29)
 Go version:   go1.7.5
 Git commit:   2878a85
 Built:        Tue Apr 11 20:55:05 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Tue Apr  4 00:37:25 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

I'm using the latest version of DFP and Swarm Listener.

I'm including the 3 stack files and the full log from start of the proxy stack to start of the vote stack which continually retries via the swarm-listener. 

Also I don't think it's a connectivity issue because when I remove the `reqPathSearch` and `reqPathReplace` labels I'm able to still use `<elb>/vote` and `<elb>/results` and see in the logs a 404 on the container I configured via the df labels.  So that part seems to be working fine.  

**proxy-stack.yaml:**
```yaml
version: '3.2'
services:

  proxy:
    image: vfarcic/docker-flow-proxy
    ports:
      - 80:80
    networks:
      - proxy
    environment:
      - LISTENER_ADDRESS=swarm-listener
      - MODE=swarm
    deploy:
      replicas: 2

  swarm-listener:
    image: vfarcic/docker-flow-swarm-listener
    networks:
      - proxy
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    environment:
      - DF_NOTIFY_CREATE_SERVICE_URL=http://proxy:8080/v1/docker-flow-proxy/reconfigure
      - DF_NOTIFY_REMOVE_SERVICE_URL=http://proxy:8080/v1/docker-flow-proxy/remove
    deploy:
      placement:
        constraints: [node.role == manager]

networks:
  proxy:
    external: true
```

**services-stack.yaml:**
```yaml
version: "3"
services:

  redis:
    image: redis:alpine
    ports:
      - "6379"
    networks:
      - service
    deploy:
      replicas: 1
      restart_policy:
        condition: on-failure

  db:
    image: postgres:9.4
    volumes:
      - db-data:/var/lib/postgresql/data
    networks:
      - service
    deploy:
      placement:
        constraints: [node.role == manager]

networks:
  service:
    external: true

volumes:
  db-data:
```

**vote-stack.yaml:**
```yaml
version: "3"
services:

  vote:
    image: dockersamples/examplevotingapp_vote:before
    networks:
      - proxy
      - service
    deploy:
      replicas: 2
      update_config:
        parallelism: 2
      restart_policy:
        condition: on-failure
      labels:
        - com.df.notify=true
        - com.df.distribute=true
        - com.df.xForwardedProto=true
        - com.df.servicePath=/votes
        - com.df.reqPathSearch='/vote/'
        - com.df.reqPathReplace='/'
        - com.df.port=80

  result:
    image: dockersamples/examplevotingapp_result:before
    networks:
      - proxy
      - service
    deploy:
      replicas: 1
      restart_policy:
        condition: on-failure
      labels:
        - com.df.notify=true
        - com.df.distribute=true
        - com.df.xForwardedProto=true
        - com.df.servicePath=/results
        - com.df.reqPathSearch='/results/'
        - com.df.reqPathReplace='/'
        - com.df.port=80

  worker:
    image: dockersamples/examplevotingapp_worker
    networks:
      - service
    deploy:
      replicas: 1
      labels: [APP=VOTING]
      restart_policy:
        condition: on-failure
        delay: 10s
        max_attempts: 3
        window: 120s
      placement:
        constraints: [node.role == manager]

networks:
  proxy:
    external: true
  service:
    external: true
```

**Proxy Log file:**
```bash
08:02:17 2017/05/01 12:02:17 Starting HAProxy
08:02:17 2017/05/01 12:02:17 Starting "Docker Flow: Proxy"
08:02:22 2017/05/01 12:02:22 Got configuration from http://swarm-listener:8080.
08:04:24 2017/05/01 12:04:24 Sending distribution request to http://10.0.0.4:8080/v1/docker-flow-proxy/reconfigure?distribute=false&port=80&serviceName=vote_result&servicePath=%2Fresults&xForwardedProto=true
08:04:24 2017/05/01 12:04:24 Creating configuration for the service vote_result
08:04:24 2017/05/01 12:04:24 Reloading the proxy
08:04:24 2017/05/01 12:04:24 Proxy config was reloaded
08:04:24 2017/05/01 12:04:24 Sending distribution request to http://10.0.0.3:8080/v1/docker-flow-proxy/reconfigure?distribute=false&port=80&serviceName=vote_result&servicePath=%2Fresults&xForwardedProto=true
08:04:34 2017/05/01 12:04:34 Creating configuration for the service vote_vote
08:04:34 2017/05/01 12:04:34 Reloading the proxy
08:04:45 2017/05/01 12:04:45 Command -f /cfg/haproxy.cfg -D -p /var/run/haproxy.pid -sf 14
08:04:45 exit status 1
08:04:45 global
08:04:45 pidfile /var/run/haproxy.pid
08:04:45 tune.ssl.default-dh-param 2048
08:04:45 # disable sslv3, prefer modern ciphers
08:04:45 ssl-default-bind-options no-sslv3
08:04:45 ssl-default-bind-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS
08:04:45 ssl-default-server-options no-sslv3
08:04:45 ssl-default-server-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS
08:04:45 defaults
08:04:45 mode http
08:04:45 balance roundrobin
08:04:45 default-server init-addr last,libc,none
08:04:45 option dontlognull
08:04:45 option dontlog-normal
08:04:45 option http-keep-alive
08:04:45 option forwardfor
08:04:45 option redispatch
08:04:45 errorfile 400 /errorfiles/400.http
08:04:45 errorfile 403 /errorfiles/403.http
08:04:45 errorfile 405 /errorfiles/405.http
08:04:45 errorfile 408 /errorfiles/408.http
08:04:45 errorfile 429 /errorfiles/429.http
08:04:45 errorfile 500 /errorfiles/500.http
08:04:45 errorfile 502 /errorfiles/502.http
08:04:45 errorfile 503 /errorfiles/503.http
08:04:45 errorfile 504 /errorfiles/504.http
08:04:45 maxconn 5000
08:04:45 timeout connect 5s
08:04:45 timeout client 20s
08:04:45 timeout server 20s
08:04:45 timeout queue 30s
08:04:45 timeout tunnel 3600s
08:04:45 timeout http-request 5s
08:04:45 timeout http-keep-alive 15s
08:04:45 frontend services
08:04:45 bind *:80
08:04:45 bind *:443
08:04:45 mode http
08:04:45 acl url_vote_result80 path_beg /results
08:04:45 use_backend vote_result-be80 if url_vote_result80
08:04:45 acl url_vote_vote80 path_beg /votes
08:04:45 use_backend vote_vote-be80 if url_vote_vote80
08:04:45 backend vote_result-be80
08:04:45 mode http
08:04:45 http-request add-header X-Forwarded-Proto https if { ssl_fc }
08:04:45 server vote_result vote_result:80 check
08:04:45 backend vote_vote-be80
08:04:45 mode http
08:04:45 http-request add-header X-Forwarded-Proto https if { ssl_fc }
08:04:45 http-request set-path %!
08:04:45 (BADINDEX) server vote_vote vote_vote:80 check
08:04:45 2017/05/01 12:04:45 Command -f /cfg/haproxy.cfg -D -p /var/run/haproxy.pid -sf 14
08:04:45 exit status 1
```

**Comment by vfarcic** [comment]
Thanks for the stacks. It'll be easy to reproduce it on my Docker for AWS cluster.

I'm not in front of a computer right now. I can go through it tonight if that's not too late.

**Comment by vfarcic** [comment]
Found the problem... It's one of those nuances...

If you put `'` around label value, Docker transforms it into encoded version. So, DFP got it and generated the line that follows in its config.

```
http-request set-path %[path,regsub(&#39;/vote/&#39;,&#39;/&#39;)]
```

Once I removed `' `, and made a few other changes, it worked.

What I did is:

* Removed `' ` from [L19](https://github.com/vfarcic/docker-flow-proxy/blob/master/issues/200/vote-stack.yaml#L19), [L20](https://github.com/vfarcic/docker-flow-proxy/blob/master/issues/200/vote-stack.yaml#L20), [L37](https://github.com/vfarcic/docker-flow-proxy/blob/master/issues/200/vote-stack.yaml#L37), and [L38](https://github.com/vfarcic/docker-flow-proxy/blob/master/issues/200/vote-stack.yaml#L38).
* Changed `vote` to `votes` in [L20](https://github.com/vfarcic/docker-flow-proxy/blob/master/issues/200/vote-stack.yaml#L20) (I guess it was a typo)
* Removed trailing slash (`/`) from [L20](https://github.com/vfarcic/docker-flow-proxy/blob/master/issues/200/vote-stack.yaml#L20) and [L38](https://github.com/vfarcic/docker-flow-proxy/blob/master/issues/200/vote-stack.yaml#L38). It would work with them as well but you'd be forced to use a trailing slash in your URLs.

Since DFP retries on a failure (for example, Docker might still be pulling a service), I removed both the proxy and vote stack, corrected the vote YAML, and deployed both stacks. It worked, kind of. Styles are still not working. They would need to be added to the proxy as well. Styles and JS are using root address (terrible practice). You'd need to add them as well. Normally, I'd inspect requests using Chrome and see which ones are missing and add them to the proxy.

Please try it out and let me know how it went. If you have any other problem, feel free to add it here (or a separate ticket if it's not related) or ping me on http://slack.devops20toolkit.com/.

**Comment by ambrons** [comment]
@vfarcic thanks I'll try your suggestions and get back to you later.

On the part of the trailing slashes, the documentation for `reqPathReplace` and `reqPathSearch` state they're regular expressions, though I assumed it was just a plan string.  So one might assume that the leading and trailing slashes are boundaries for the regular expression.  Beyond the fault in my configuration you might want to update the documentation in that section.

http://proxy.dockerflow.com/usage/
**reqPathReplace**
> A regular expression to apply the modification. If specified, reqPathSearch needs to be set as well.
Example: /demo/

**reqPathSearch**
> A regular expression to search the content to be replaced. If specified, reqPathReplace needs to be set as well.
Example: /something/

I hadn't made it far enough with the vote app to realize the root url.  That's unfortunately.  I was using it as a prime example to route-based proxying and was going to use it for name-based proxying as well (I've done this before with another project and works great).

BTW ready your book great stuff!  Had a couple of suggestions, but this isn't the forum for it.

**Comment by vfarcic** [comment]
`reqPathReplace` and `reqPathSearch` are regular expressions. Their values are, but the full expression is assembled by DFP and written in HAProxy format. So, no need to slashes unless they are part of the path. Now when you say, I realized that it might really be misleading. I would probably interpret it the same as you (if it was someone else's project). I'll update the docs to make them clearer.

In my case, whenever there are multiple paths I tend to deploy services under different domains. It simplifies everything a lot. Actually, I tend to deploy all UI based services in different domains. It's not only about simplifying configs but also it's more user-friendly to open something like `prometheus.acme.com` than `acme.com/prometheus`. However, that's my personal preference for production and others probably follow different patterns.

Please let me know if you have any other question related to this issue or I can close it.

P.S. Thanks for the words of praise for the book. You should put them in Amazon.com. For us, self-published authors, that's the only marketing we have.

**Comment by ambrons** [comment]
So I was able to get things working with your suggested changes.  I probably had it working the first time too, but the typeo got me I'm sure.

One thing about the single-quotes and documentation.  The original link I posted: 
http://proxy.dockerflow.com/swarm-mode-auto/#rewriting-paths

has the following command-line snippet.  I didn't test if a `docker create service` works differently or not then the yaml stack file, but something that might also need updating:
```
docker service create --name go-demo \
    -e DB=go-demo-db \
    --network go-demo \
    --network proxy \
    --label com.df.notify=true \
    --label com.df.distribute=true \
    --label com.df.servicePath=/something \
    --label com.df.port=8080 \
    --label com.df.reqPathSearch='/something/' \
    --label com.df.reqPathReplace='/demo/' \
    --replicas 3 \
    vfarcic/go-demo
```

Sucks about the vote app I choose to use as an example.  Might just write a small node app myself or fork that other project, though seems like for kicks the guy wrote each service in another language.

Yes I agree.  The production swarm I'm managing I setup subdomains for each service and am using the `serviceDomain` label to manage routing.  However I'm writing some documentation and wanted to show the other side of the coin for those that choose to do so.  So it's for example purposes only.

I bought your book from leanpub, but I'll head over to amazon and give it a great review.
Had a lot about the monitoring section as an enhancement, but this really isn't the forum for it.  I'll hit you up on your slack.

I'm good for closing this ticket, but in honestly I highjacked this issue, so not sure if @patran needs this open for any reason.
